### PR TITLE
Get latest counter before attempting a take to ensure take succeeds

### PR DESCRIPTION
**Pull request by zengyu714**
*Issue #, if available:*
- https://github.com/awslabs/amazon-kinesis-client/pull/765/

*Description of changes:*
- Besides from committing the PR [#765](https://github.com/awslabs/amazon-kinesis-client/pull/765/) `Get latest counter before attempting a take to ensure take succeeds` raised by @Renjuju, also addressed comments left by @ashwing.

*Previous Changes*
https://github.com/ashwing/amazon-kinesis-client/pull/94
https://github.com/awslabs/amazon-kinesis-client/pull/765/

*Verification*
1. When artificially introduce delay while fetching leases, kcl instance does attempt to refresh the lease and succeed stealing the lease
2. Property `isMarkedForSteal` is updated for each lease in each cycle of renewing leases
 

See logs here
```
2022-01-03 15:02:37,855 [LeaseCoordinator-0000] INFO  s.a.k.l.dynamodb.DynamoDBLeaseTaker [NONE] - [Functional Test] Check isMarkedForLeaseSteal before assign true values: [false], which is shouldn't be true 
2022-01-03 15:02:37,856 [LeaseCoordinator-0000] INFO  s.a.k.l.dynamodb.DynamoDBLeaseTaker [NONE] - [Functional Test] leases with isMarkedForLeaseSteal==true should be different with last cycle: [true] 
2022-01-03 15:02:37,856 [LeaseCoordinator-0000] INFO  s.a.k.l.dynamodb.DynamoDBLeaseTaker [NONE] - [Functional Test] leaseKeysToStealFromLastCycleSet: [shardId-000000000014] and leaseKeysToStealFromCurrentCycleSet: [shardId-000000000010] 
2022-01-03 15:02:37,856 [LeaseCoordinator-0000] INFO  s.a.k.l.dynamodb.DynamoDBLeaseTaker [NONE] - Worker mac-1 needed 9 leases but none were expired, so it will steal lease shardId-000000000010 from mac-2 
2022-01-03 15:02:37,856 [LeaseCoordinator-0000] INFO  s.a.k.l.dynamodb.DynamoDBLeaseTaker [NONE] - Worker mac-1 saw 20 total leases, 0 available leases, 2 workers. Target is 10 leases, I have 1 leases, I will take 1 leases 
2022-01-03 15:02:37,856 [LeaseCoordinator-0000] INFO  s.a.k.l.dynamodb.DynamoDBLeaseTaker [NONE] - [Functional Test] Updating stale leases 
2022-01-03 15:02:37,856 [LeaseCoordinator-0000] INFO  s.a.k.l.dynamodb.DynamoDBLeaseTaker [NONE] - [Functional Test] Attempt to refresh the lease as expected 
2022-01-03 15:02:37,991 [LeaseCoordinator-0000] INFO  s.a.k.l.dynamodb.DynamoDBLeaseTaker [NONE] - Worker mac-1 successfully took 1 leases: shardId-000000000010 
2022-01-03 15:02:38,404 [KinesisTester-0000] INFO  s.a.kinesis.coordinator.Scheduler [NONE] - Created new shardConsumer for : ShardInfo(streamIdentifierSerOpt=Optional.empty, shardId=shardId-000000000010, concurrencyToken=cc7f5fd6-471b-4a2a-b59e-dc108f19118b, parentShardIds=[], checkpoint={SequenceNumber: 49625136119961813246747861896741176352940323142511886498,SubsequenceNumber: 0}) 
```

For all changes made for testing, see here: https://gist.github.com/zengyu714/ef2b46b051a97b4e43184948f25d7a93/revisions?diff=split
 - Note the artificial delay is introduced by adding lines before [here](https://github.com/awslabs/amazon-kinesis-client/blob/b48de6d1bfea2f4139c42fd2772a9219b26c1cc1/amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java#L176-L177)
```
                if (Objects.equals(workerIdentifier, "mac-1")){
                    updateAllLeasesTotalTimeMillis += leaseRenewalIntervalMillis;
                }
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java [hunk 1]
```diff
@@ -233,6 +242,32 @@ public DynamoDBLeaseTaker withMaxLeasesToStealAtOneTime(int maxLeasesToStealAtOn
...
+     * @return set of leases to take.
+     */
+    private Set<Lease> updateStaleLeasesWithLatestState(long updateAllLeasesEndTime,
+                                                        Set<Lease> leasesToTake) {
+        if (updateAllLeasesEndTime > leaseRenewalIntervalMillis * RENEWAL_SLACK_PERCENTAGE) {
+            leasesToTake = leasesToTake.stream().map(lease -> {
+                if (lease.isMarkedForLeaseSteal()) {
+                    try {
+                        return leaseRefresher.getLease(lease.leaseKey());
+                    } catch (DependencyException | InvalidStateException | ProvisionedThroughputException e) {
```
Comment (avahuang0429) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java:262]: https://github.com/awslabs/amazon-kinesis-client/pull/765#discussion_r574736673

this needs to be addressed

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/CommonCalculations.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+
+    /**
+     * Convenience method for calculating renewer intervals in milliseconds.
+     *
+     * @param leaseDurationMillis Duration of a lease
+     * @param epsilonMillis       Allow for some variance when calculating lease expirations
+     */
+    public static long getRenewerTakerIntervalMillis(long leaseDurationMillis, long epsilonMillis) {
+        return leaseDurationMillis / 3 - epsilonMillis;
```
Comment (avahuang0429) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/CommonCalculations.java:29]: is this essentially veryOldLeaseDurationNanosMultiplier?

**Comment by avahuang0429** [comment]
Also can you check why travis ci is complaining? might be transient but double confirm by rerunning

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java (hunk 1 above)
Comment (zengyu714) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java:262]: i think `while refreshing the stale lease` explains why we get this exception, following the suggestion from Ashwing `Like while we tried to update the stale leases` in #765 [(comment)](https://github.com/awslabs/amazon-kinesis-client/pull/765#discussion_r574736673). And the line `defaulting to existing lease` explains the current status... do you want more context on the failure?

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/CommonCalculations.java (hunk 2 above)
Comment (zengyu714) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/CommonCalculations.java:29]: yep, being extracted into a static function

**Comment by zengyu714** [comment]
> Also can you check why travis ci is complaining? might be transient but double confirm by rerunning

@avahuang0429 the new run succeeded: https://app.travis-ci.com/github/awslabs/amazon-kinesis-client/builds/243951865

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/CommonCalculations.java (hunk 2 above)
Comment (avahuang0429) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/CommonCalculations.java:29]: can you use veryOldLeaseDurationNanosMultiplier instead?

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java (hunk 1 above)
Comment (avahuang0429) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java:262]: "Unable to retrieve the current lease while refreshing the stale lease" this means that we know for a fact that the leases being refreshed are stale while in fact they can be just the same as in DDB. This is not accurate. You can explain in comment naming different cases we run into exceptions. 

"Like while we tried to update the stale leases" <- i am not sure this is the right reason why we would run into such exception. He might have confused it with a conditional update; while this is in fact just a get ddb request. I am guessing his intention is we explain how we would run into dependencyException and invalidStateException

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/CommonCalculations.java (hunk 2 above)
Comment (zengyu714) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/common/CommonCalculations.java:29]: it's not from my commit, do you have strong preference about it?

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java (hunk 1 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java:262]: Here we are trying to update the local copy of the leases with latest info in the event of leases going stale. Note that we already fetched the leases from ddb table just a while ago, but now we want to get their latest state in order to successfully steal the leases. Having a message like "Failed to fetch latest state of the lease <leaseKey> that needs to be stolen"

Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java:262]: can you add leasekey to the msg

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java [hunk 3]
```diff
@@ -233,6 +243,33 @@ public DynamoDBLeaseTaker withMaxLeasesToStealAtOneTime(int maxLeasesToStealAtOn
...
+                        log.warn("Failed to fetch latest state of the lease {} that needs to be stolen, "
```
Comment (zengyu714) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java:263]: added `leaseKey` in the logline

**Comment by ashwing** [comment]
Thanks I need to go over the PR one more time. Can you do a functional test for this change and update the CR?

Yu Zeng 10:33 AM
sure, just to confirm tha: does the functional test mean Testing KCL in the local?

Ashwin Giridharan 10:34 AM
yeah basically you need to run two instances locally, with different worker identifier
10:34
so that one can steal leases from another (edited)

10:35
when you artificially introduce delay while fetching leases, the worker should attempt to refresh the lease (as per this PR
) and succeed stealing the lease
10:36
Also we should verify new leases are created every lease renewal i.e we should use isMarkedForSteal for only one cycle. (edited)

Yu Zeng 10:37 AM
got it, thanks!

**Comment by zengyu714** [comment]
> Thanks I need to go over the PR one more time. Can you do a functional test for this change and update the CR?
> 
> Yu Zeng 10:33 AM sure, just to confirm tha: does the functional test mean Testing KCL in the local?
> 
> Ashwin Giridharan 10:34 AM yeah basically you need to run two instances locally, with different worker identifier 10:34 so that one can steal leases from another (edited)
> 
> 10:35 when you artificially introduce delay while fetching leases, the worker should attempt to refresh the lease (as per this PR ) and succeed stealing the lease 10:36 Also we should verify new leases are created every lease renewal i.e we should use isMarkedForSteal for only one cycle. (edited)
> 
> Yu Zeng 10:37 AM got it, thanks!

Verification section is added in the description with example logs: https://github.com/awslabs/amazon-kinesis-client/pull/886#issue-1088012043

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java (hunk 1 above)
Comment (zengyu714) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseTaker.java:262]: `leaseKey` is added to the logline

**Review by ashwing — APPROVED** [review]
lgtm

**Pull request merged**